### PR TITLE
Guard clause in RecursionGuard and OmitOnRecursionBehavior constructors

### PR DESCRIPTION
**Pull request by Pvlerick**
Guard clause for the recursionDepth parameter that should be strictly higher than zero.
Unit tests added in both cases.

Addressing #337

### CODE REVIEW

**Comment by moodmosaic** [comment]
I've just verified the code in this Pull Request, with the code sample provided by @dcastro [here](https://github.com/AutoFixture/AutoFixture/issues/337#issue-54626803), and seems to be working as expected now:

```csharp
var fixture = new Fixture();
fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList()
    .ForEach(b => fixture.Behaviors.Remove(b));
fixture.Behaviors.Add(new OmitOnRecursionBehavior(0));

fixture.Create<string>();
```
Output:

```
Test 'Ploeh.AutoFixtureUnitTest.CollectionFillerTest.Test' failed
    : System.ArgumentOutOfRangeException 
    : Specified argument was out of the range of valid values.
Parameter name: recursionDepth
    OmitOnRecursionBehavior.cs(33,0)
        : at Ploeh.AutoFixture.OmitOnRecursionBehavior..ctor(Int32 recursionDepth)
    CollectionFillerTest.cs(24,0)
        : at Ploeh.AutoFixtureUnitTest.CollectionFillerTest.Test()
```

File: Src/AutoFixtureUnitTest/OmitOnRecursionBehaviorTest.cs [hunk 1]
```diff
@@ -77,5 +77,15 @@ public void TransformResultCorrectlyDecoratesInput()
...
             // Teardown
         }
+
+        [Fact]
+        public void ConstructorWithRecusionDepthLowerThanOneThrows()
+        {
+            // Fixture setup
+            // Exercise system and verify outcome
+            Assert.Throws<ArgumentOutOfRangeException>(() =>
+                new OmitOnRecursionBehavior(0));
```
Comment (moodmosaic) [inline Src/AutoFixtureUnitTest/OmitOnRecursionBehaviorTest.cs]: Wouldn't it be more clear, that recursion depth must be lower than `1`, if we'd parameterize the test? (e.g. `0`, `-7`, `-21`)

Comment (Pvlerick) [inline Src/AutoFixtureUnitTest/OmitOnRecursionBehaviorTest.cs]: Didn't think about it, indeed - I'll turn it into a Theory.

File: Src/AutoFixture/Kernel/RecursionGuard.cs [hunk 2]
```diff
@@ -173,6 +174,9 @@ public RecursionGuard(ISpecimenBuilder builder, IEqualityComparer comparer)
             if (comparer == null)
                 throw new ArgumentNullException("comparer");
 
+            if (recursionDepth < 1)
+                throw new ArgumentOutOfRangeException("recursionDepth");
```
Comment (ploeh) [inline Src/AutoFixture/Kernel/RecursionGuard.cs]: Would it be more helpful to the user if the exception message explains what is the acceptable range?

File: Src/AutoFixture/OmitOnRecursionBehavior.cs [hunk 3]
```diff
@@ -26,8 +26,12 @@ public OmitOnRecursionBehavior() : this(DefaultRecursionDepth)
         /// Initializes new instance of the <see cref="OmitOnRecursionBehavior" /> class with specific recursion depth.
         /// </summary>
         /// <param name="recursionDepth">The recursion depth at which the request will be omitted.</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">recursionDepth is less than zero.</exception>
         public OmitOnRecursionBehavior(int recursionDepth)
         {
+            if (recursionDepth < 1)
+                throw new ArgumentOutOfRangeException("recursionDepth");
```
Comment (ploeh) [inline Src/AutoFixture/OmitOnRecursionBehavior.cs]: Would it be more helpful to the user if the exception message explains what is the acceptable range?

**Comment by Pvlerick** [comment]
Thanks for your reviews; should be good now.

**Comment by ploeh** [comment]
Thank you for your contribution! It's now live as [AutoFixture 3.30.5](http://www.nuget.org/packages/AutoFixture/3.30.5).

**Pull request merged**